File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
  const bridge = Array(bridge_length).fill(0);
  let time = 0;
  let currentWeight = 0;

  while (bridge.length > 0) {
    time++;
    currentWeight -= bridge.shift();

    if (truck_weights.length > 0) {
      if (currentWeight + truck_weights[0] <= weight) {
        const truck = truck_weights.shift();
        bridge.push(truck);
        currentWeight += truck;
      } else {
        bridge.push(0);
      }
    }

    if (truck_weights.length === 0 && currentWeight === 0) break;
  }

  return time;
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Queue<Integer> bridge = new LinkedList<>();

        // 다리 초기화
        for (int i = 0; i < bridge_length; i++) {
            bridge.offer(0);
        }

        int time = 0;
        int currentWeight = 0;
        int index = 0;

        while (!bridge.isEmpty()) {
            time++;

            // 다리에서 트럭(또는 빈칸) 하나 빠짐
            currentWeight -= bridge.poll();

            // 아직 대기 중인 트럭이 있다면
            if (index < truck_weights.length) {
                if (currentWeight + truck_weights[index] <= weight) {
                    bridge.offer(truck_weights[index]);
                    currentWeight += truck_weights[index];
                    index++;
                } else {
                    bridge.offer(0);
                }
            }

            // 모든 트럭이 다리를 건넜으면 종료
            if (index == truck_weights.length && currentWeight == 0) {
                break;
            }
        }

        return time;
    }
}